src/main/java/com/github/checkstyle/regression/configuration/ConfigGenerator.java

```diff
@@ -0,0 +1,187 @@
+    }
+
+    /**
+     * Creates a new {@link Document} instance from the {@code BASE_CONFIG}.
+     * @return a new {@link Document} instance
+     */
+    private static Document createXmlDocument() {
+        final Document document;
+
+        try {
+            final File baseConfig = new File("src/main/resources/com/github/checkstyle/regression/"
+                    + "configuration/base_config.xml");
+            document = DocumentBuilderFactory.newInstance().newDocumentBuilder()
+                    .parse(baseConfig);
+        }
+        catch (ParserConfigurationException | SAXException | IOException ex) {
+            throw new IllegalStateException("cannot instantiate Document instance", ex);
+        }
+
+        return document;
+    }
+
+    /**
+     * Creates a new {@link Transformer} instance and do initialization.
+     * @return a new {@link Transformer} instance
+     */
+    private static Transformer createXmlTransformer() {
+        final Transformer transformer;
+
+        try {
+            transformer = TransformerFactory.newInstance().newTransformer();
+            transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+            transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
+            transformer.setOutputProperty(OutputKeys.DOCTYPE_PUBLIC, DOCTYPE_PUBLIC);
+            transformer.setOutputProperty(OutputKeys.DOCTYPE_SYSTEM, DOCTYPE_SYSTEM);
+        }
+        catch (TransformerConfigurationException ex) {
+            throw new IllegalStateException("cannot instantiate Transformer instance", ex);
+        }
+
+        return transformer;
+    }
+
+    /**
+     * Generates the plain text of the config XML from the given module infos.
+     * @param moduleInfos the given module infos
+     * @return the generated plain text of the config
+     * @throws TransformerException failure of transforming the XML document
+     */
+    public static String generateConfigText(List<ModuleInfo> moduleInfos)
+            throws TransformerException {
+        final Document document = createXmlDocument();
+
+        final Node checkerNode = getCheckerModuleNode(document);
+        final Node treeWalkerNode = getTreeWalkerModuleNode(document);
+        for (ModuleInfo moduleInfo : moduleInfos) {
+            final Node moduleNode = createModuleNode(document, moduleInfo);
+            final String parent = moduleInfo.moduleExtractInfo().parent();
+            if (PARENT_CHECKER.equals(parent)) {
+                checkerNode.appendChild(moduleNode);
+            }
+            else if (PARENT_TREE_WALKER.equals(parent)) {
+                treeWalkerNode.appendChild(moduleNode);
+            }
+        }
+
+        final Transformer transformer = createXmlTransformer();
+        final StringWriter writer = new StringWriter();
+        transformer.transform(new DOMSource(document), new StreamResult(writer));
+        return writer.getBuffer().toString().replaceAll("\r\n", "\n");
```

Comment:
> .replaceAll("\r\n", "\n");

Does this really need to be in production code? We are no longer doing any processing on this and Checkstyle/XML doesn't care about whitespace.
I assume this was added just to make tests pass, so why don't we keep it in test code.